frontend/deb/debroot.go

```diff
@@ -0,0 +1,438 @@
+//  2. Sources with certain characters in the name had to be changed, so we need
+//     to bring those back.
+//
+// This is called from `debian/rules` after the source tarball has been extracted.
+func fixupSources(spec *dalec.Spec) []byte {
+	buf := bytes.NewBuffer(nil)
+	writeScriptHeader(buf)
+
+	// now, we need to find all the sources that are file-backed and fix them up
+	for name, src := range spec.Sources {
+		dirName := sanitizeSourceKey(name)
+
+		if dalec.SourceIsDir(src) {
+			if dirName == name {
+				continue
+			}
+			fmt.Fprintf(buf, "mv '%s' '%s'\n", dirName, name)
+			continue
+		}
+
+		fmt.Fprintf(buf, "mv '%s/%s' '%s.dalec.tmp' || (ls -lh %q; exit 2)\n", dirName, name, name, dirName)
+		fmt.Fprintf(buf, "rm -rf '%s'\n", dirName)
+		fmt.Fprintf(buf, "mv '%s.dalec.tmp' '%s'\n", name, name)
+		fmt.Fprintln(buf)
+	}
+
+	return buf.Bytes()
+}
+
+func writeScriptHeader(buf io.Writer) {
+	fmt.Fprintln(buf, "#!/usr/bin/env sh")
+	fmt.Fprintln(buf)
+
+	fmt.Fprintln(buf, "set -ex")
+}
+
+func createPatchScript(spec *dalec.Spec) []byte {
+	buf := bytes.NewBuffer(nil)
+
+	writeScriptHeader(buf)
+
+	for name, patches := range spec.Patches {
+		for _, patch := range patches {
+			p := filepath.Join("${DEBIAN_DIR:=debian}/dalec/patches", name, patch.Source)
+			fmt.Fprintf(buf, "patch -d %q -p%d -s < %q\n", name, *patch.Strip, p)
+		}
+	}
+
+	return buf.Bytes()
+}
+
+func createBuildScript(spec *dalec.Spec) []byte {
+	buf := bytes.NewBuffer(nil)
+	writeScriptHeader(buf)
+
+	sorted := dalec.SortMapKeys(spec.Build.Env)
+	for _, k := range sorted {
+		v := spec.Build.Env[k]
+		fmt.Fprintf(buf, "export %q=%q\n", k, v)
+	}
+
+	for _, step := range spec.Build.Steps {
+		fmt.Fprintln(buf)
+		fmt.Fprintln(buf, "(")
+
+		sorted := dalec.SortMapKeys(step.Env)
+		for _, k := range sorted {
+			v := step.Env[k]
+			fmt.Fprintf(buf, "	export %q=%q\n", k, v)
+		}
+
+		fmt.Fprintln(buf, step.Command)
+		fmt.Fprintln(buf, ")")
+	}
+
+	return buf.Bytes()
+}
+
+func createInstallScripts(worker llb.State, spec *dalec.Spec, dir string) []llb.State {
+	if spec.Artifacts.IsEmpty() {
+		return nil
+	}
+
+	states := make([]llb.State, 0, len(spec.Artifacts.Binaries)+len(spec.Artifacts.Manpages))
+	base := llb.Scratch().File(llb.Mkdir(dir, 0o755, llb.WithParents(true)))
+
+	installBuf := bytes.NewBuffer(nil)
+	writeInstallHeader := sync.OnceFunc(func() {
+		fmt.Fprintln(installBuf, "#!/usr/bin/dh-exec")
+		fmt.Fprintln(installBuf)
+	})
+
+	writeInstall := func(src, dir, name string) {
+		writeInstallHeader()
+
+		if filepath.Base(src) != name {
+			fmt.Fprintln(installBuf, src, "=>", filepath.Join(dir, name))
+			return
+		}
+
+		fmt.Fprintln(installBuf, src, dir+"/")
+	}
+
+	if len(spec.Artifacts.Binaries) > 0 {
+		sorted := dalec.SortMapKeys(spec.Artifacts.Binaries)
+		for _, key := range sorted {
+			cfg := spec.Artifacts.Binaries[key]
+			writeInstall(key, filepath.Join("/usr/bin", cfg.SubPath), cfg.ResolveName(key))
```

Comment:
Should `/usr/bin` be hardcoded here?

Reply:
Do you have something else in mind? I don't think there's any macro or anything to go with here.

Reply:
Yeah, I suppose I was wondering if there was a macro or some kind of debhelper command which would return configured installation paths. But it's looking like that really doesn't exist in which case I guess hardcoding according to the proper conventions may be the best thing we can do